src/lambda.ts

```diff
@@ -0,0 +1,134 @@
+import type {
+  APIGatewayEvent,
+  APIGatewayProxyEventV2,
+  APIGatewayProxyResult,
+  Context,
+} from "aws-lambda";
+import {
+  InngestCommHandler,
+  ServeHandler,
+} from "./components/InngestCommHandler";
+import { headerKeys, queryKeys } from "./helpers/consts";
+import { allProcessEnv } from "./helpers/env";
+
+/**
+ * With AWS Lambda, serve and register any declared functions with Inngest,
+ * making them available to be triggered by events.
+ *
+ * @example
+ *
+ * ```ts
+ * import { Inngest } from "inngest";
+ * import { serve } from "inngest/lambda";
+ *
+ * const inngest = new Inngest({ name: "My Lambda App" });
+ *
+ * const fn = inngest.createFunction(
+ *   { name: "Hello World" },
+ *   { event: "test/hello.world" },
+ *   async ({ event }) => {
+ *     return "Hello World";
+ *   }
+ * );
+ *
+ * export const handler = serve(inngest, [fn]);
+ * ```
+ *
+ * @public
+ */
+export const serve: ServeHandler = (nameOrInngest, fns, opts) => {
+  const handler = new InngestCommHandler(
+    "aws-lambda",
+    nameOrInngest,
+    fns,
+    { ...opts },
+    (event: APIGatewayEvent | APIGatewayProxyEventV2, _context: Context) => {
+      /**
+       * Try to handle multiple incoming event types, as Lambda can have many
+       * triggers.
+       *
+       * This still doesn't handle all cases, but it's a start.
+       */
+      const eventIsV2 = ((
+        ev: APIGatewayEvent | APIGatewayProxyEventV2
+      ): ev is APIGatewayProxyEventV2 => {
+        return (ev as APIGatewayProxyEventV2).version === "2.0";
+      })(event);
+
+      const method = eventIsV2
+        ? event.requestContext.http.method
+        : event.httpMethod;
+      const path = eventIsV2 ? event.requestContext.http.path : event.path;
+
+      const env = allProcessEnv();
+      const proto = event.headers["x-forwarded-proto"] || "https";
+
+      const url = new URL(path, `${proto}://${event.headers.host || ""}`);
+
+      const isProduction =
+        env.CONTEXT === "production" ||
+        env.ENVIRONMENT === "production" ||
+        env.NODE_ENV?.startsWith("prod") ||
+        false;
+
+      return {
+        register: () => {
+          if (method === "PUT") {
```

Comment:
I'm curious, why does every handler need to have these conditional statements within each handler function? Could we instead pass the `method` as a string to the return statement like `register` and `run`? it feels like it's duplicative in these handlers.  i.e.

```ts
// for this example
return {
  method,
  register : () => {...
}
// for other handlers that use "req"
return {
  method: req.method,
  register : () => {...
}
```

Theoretically, we could do this for a few key standard fields like `method`, `env`, `isProduction`, `url` too

Reply:
@djfarrelly Yeah, a tiny refactor that's been on the list is to generically provide `env`, `isProduction`, and `url` alongside the methods as they're required for every call; just haven't done that refactor in this PR.

We don't explicitly provide `method` as it's technically an implementation detail - e.g. it may be that a particular platform converts every request to `GET`, or maybe it never exposes the method at all. Instead, we use the `register`/`run`/`view` functions to decide what to do as it gives us a few benefits over just returning a string:

- It's easier to see what cases need to be supported, rather than having to know the internals and therefore which method to pass if you need to create one on your own.

- We can use any detail (not just HTTP method) to decide what action we should take, so it's super flexible for different platforms.

- If we add a new way to interact with the SDK, let's say `tunnel`, we could add it as [a new `HandlerAction`](https://github.com/inngest/inngest-js/blob/main/src/components/InngestCommHandler.ts#L1039-L1080) and TypeScript will enforce that all handlers now account for that action.

- As with the above point, when writing a new serve handler, the use of this pattern forces the developer to account for every single action required by the SDK, so even someone writing a new handler for their custom framework is guided through the process rather than having to understand how Inngest works.